File: src/main/java/com/codingforcookies/betterrecords/ModConfig.java
package com.codingforcookies.betterrecords;

import net.minecraftforge.common.config.Config;
import net.minecraftforge.common.config.ConfigManager;
import net.minecraftforge.fml.client.event.ConfigChangedEvent;
import net.minecraftforge.fml.common.Mod;
import net.minecraftforge.fml.common.eventhandler.SubscribeEvent;

/**
 * Config file for BetterRecords
 *
 * For now, this is written in Java, since the annotation
 * based config system does not play nice with kotlin.
 */
@Config(modid = ConstantsKt.ID)
@Config.LangKey("betterrecords.config.title")
public class ModConfig {

    @Config.Comment("Maximum speaker radius")
    @Config.RangeInt(min = -1, max = 1000)
    public static int maxSpeakerRadius = -1;

    @Config.Comment("Client-Specific config settings")
    @Config.LangKey("betterrecords.config.client.title")
    public static Client client = new Client();
    public static class Client {

        @Config.Comment("Should the mod download songs from the internet?")
        public boolean downloadSongs = true;

        @Config.Comment("Enable developer mode")
        public boolean devMode = false;

        @Config.Comment("Max file size to download (in megabytes)")
        @Config.RangeInt(min = 1, max = 1000)
        public int downloadMax = 10;

        @Config.Comment("Play Songs while downloading\nFor those with fast internet!")
        public boolean playWhileDownloading = false;

        @Config.RangeInt(min = 256, max = 2048)
        public int streamBuffer = 1024;

        @Config.Comment("Should radio be streamed")
        public boolean streamRadio = true;

        @Config.Comment("Intensity of lights")
        @Config.RangeInt(min = -1, max = 3)
        public int flashMode;
    }

    public static void update() {
        ConfigManager.sync(ConstantsKt.ID, Config.Type.INSTANCE);
    }

    @Mod.EventBusSubscriber
    private static class EventHandler {

        /**
         * Inject the new values and save to the config file when the config has been changed from the GUI.
         *
         * @param event The event
         */
        @SubscribeEvent
        public static void onConfigChanged(ConfigChangedEvent.OnConfigChangedEvent event) {
            if (event.getModID().equals(ConstantsKt.ID)) {
                ModConfig.update();
            }
        }
    }
}

File: src/main/kotlin/com/codingforcookies/betterrecords/ModConfig.kt
package com.codingforcookies.betterrecords

import net.minecraftforge.common.config.Config
import net.minecraftforge.common.config.ConfigManager
import net.minecraftforge.fml.client.event.ConfigChangedEvent
import net.minecraftforge.fml.common.Mod
import net.minecraftforge.fml.common.eventhandler.SubscribeEvent

/**
 * Config file for BetterRecords
 *
 * As of right now, having this in Kotlin creates an unnecessary instance{} block in the config.
 * I don't really care.
 */
@Config(modid = ID)
@Config.LangKey("betterrecords.config.title")
object ModConfig {

    @JvmField
    @Config.Comment("Maximum speaker radius")
    @Config.RangeInt(min = -1, max = 1000)
    var maxSpeakerRadius = -1

    @JvmField
    @Config.Comment("Client-Specific config settings")
    @Config.LangKey("betterrecords.config.client.title")
    var client = Client()
    class Client {

        @JvmField
        @Config.Comment("Should the mod download songs from the internet?")
        var downloadSongs = true

        @JvmField
        @Config.Comment("Enable developer mode")
        var devMode = false
        @JvmField
        @Config.Comment("Max file size to download (in megabytes)")
        @Config.RangeInt(min = 1, max = 1000)
        var downloadMax = 10

        @JvmField
        @Config.Comment("Play Songs while downloading\nFor those with fast internet!")
        var playWhileDownloading = false

        @JvmField
        @Config.RangeInt(min = 256, max = 2048)
        var streamBuffer = 1024

        @JvmField
        @Config.Comment("Should radio be streamed")
        var streamRadio = true

        @JvmField
        @Config.Comment("Intensity of lights")
        @Config.RangeInt(min = -1, max = 3)
        var flashMode = -1

        @JvmField
        @Config.Comment("Should the mod's built in libraries be loaded")
        var loadDefaultLibraries = true
    }

    fun update() {
        ConfigManager.sync(ID, Config.Type.INSTANCE)
    }

    @Mod.EventBusSubscriber
    private object EventHandler {

        /**
         * Inject the new values and save to the config file when the config has been changed from the GUI.
         */
        @JvmStatic
        @SubscribeEvent
        fun onConfigChanged(event: ConfigChangedEvent.OnConfigChangedEvent) {
            if (event.modID == ID) {
                ModConfig.update()
            }
        }
    }
}
